### ISSUE TITLE
comments

### ISSUE THREAD

**Issue by mivanit** [bguevel/LLMsfinal#1]
- you have a writeup.txt, but no README.md -- a .txt file does not render nicely, consider writing your final report in either markdown (github renders it, check on the website) or latex (be sure to upload the compiled pdf, and all figures/files required to compile it)
- you are missing a file map, that tells me what the various .py files do
- `dataset.py` is confusing to me. you will have to explain the dataset generation process, what the data looks like, and what the proofs actually mean
- don't commit `__pycache__`, delete it from the repo and use a `.gitignore`
- I'm not sure what `eval.py` does

Overall, you will need to be more thorough in your writeup, and explain things in a more accessible way. I do not know much about theorem proving, and will need some examples of correct proofs, LLM generated proofs, and your actual evaluation methodology to be able to understand what exactly is happening.

**Comment by mivanit** [comment]
# messy notes from our office hours chat


you want stacked embeddings that add *extra* information, something like this:
```
( A  OR ( B  AND ( X  AND Y  ) OR C  ) ) # raw text
0 10 2  0 11 3   0 12 3   13 1 2  14 1 1 # tok embeds
1 1  1  2 2  2   3 3  3   3  3 2  2  2 1 # a "depth" embedding
```

> Note: once you write this `compute_depth` function, try probing the original model (without depth embeddings) to see if it represents the depth.




```python
embed_layer_toks = nn.Embedding(d_vocab, d_model)
embed_layer_pos = nn.Embedding(max_n_context, d_model)
embed_layer_depth = nn.Embedding(max_statement_depth, d_model)

# your code should generate text that we can just split by spaces
text: str = "( A  OR ( B  AND ( X  AND Y  ) OR C  ) ) -> ..."
tokens: list[str] = text.split(" ")

text_embed: Tensor = (
  embed_layer_toks(tokenizer.encode(tokens)) # Tensor of n_ctx, d_model
  + embed_layer_pos(tensor.arange(len(tokens)) # positional embeds
  + embed_layer_depth(compute_depth(tokens)) # one model with your extra embed, a control without it
  # add more kinds of embeddings that contain information extracted from the AST
  # maybe: LHS vs RHS embedding
  # maybe: 1 if a variable, 2 if a paren, 3 if a logical statement?
)

def compute_depth(toks: list[str]):
   # something to do with counting parens
   ...
```






```

0 1 2 3 4 56 7

tokens.split(" ")

LPAREN VAR_1 AND VAR_2 RPAREN IMPLIES VAR_3
0      1     2   3     4      5       7
```


# most critical llm architecture pieces:

- RMSnorm for normalization (read nanogpt implementation for *where* to put these)
- AdamW for optimizer (dont use plain SGD)
- not sure wrt activation function (not too important)
- RoPE would be better, but you're already messing w/ embeddings, so stick to standard learned

**Comment by mivanit** [comment]
a similar project: https://github.com/sawyerfallen/math598finalproj/
